localstack-core/localstack/services/stepfunctions/asl/eval/event/execution_logging.py

```diff
@@ -0,0 +1,225 @@
+        if not log_group_name:
+            return None
+
+        include_execution_data = logging_configuration["includeExecutionData"]
+
+        return CloudWatchLoggingConfiguration(
+            state_machine_arn=state_machine_arn,
+            log_level=log_level,
+            log_group_name=log_group_name,
+            include_execution_data=include_execution_data,
+        )
+
+    def validate(self) -> None:
+        # Asserts that the logging configuration can be used for logging.
+        logs_client = connect_to().logs
+        try:
+            logs_client.create_log_stream(
+                logGroupName=self.log_group_name, logStreamName=VALIDATION_LOG_STREAM_NAME
+            )
+        except ClientError as error:
+            error_code = error.response["Error"]["Code"]
+            if error_code != "ResourceAlreadyExistsException":
+                raise InvalidLoggingConfiguration(
+                    "Invalid Logging Configuration: Log Destination not found."
+                )
+
+
+class HistoryLog(TypedDict):
+    id: str
+    previous_event_id: str
+    event_timestamp: datetime
+    type: HistoryEventType
+    execution_arn: LongArn
+    details: NotRequired[ExecutionEventLogDetails]
+
+
+class CloudWatchLoggingSession:
+    execution_arn: Final[LongArn]
+    configuration: Final[CloudWatchLoggingConfiguration]
+    _logs_client: Final[BaseClient]
+    _setup_failed: bool
+
+    def __init__(self, execution_arn: LongArn, configuration: CloudWatchLoggingConfiguration):
+        self.execution_arn = execution_arn
+        self.configuration = configuration
+        self._logs_client = connect_to().logs
+        self._setup_failed = True
+
+    def log_level_filter(self, history_event_type: HistoryEventType) -> bool:
+        # Checks whether the history event type should be logged in this session.
+        return is_event_in_context(
+            log_level=self.configuration.log_level, history_event_type=history_event_type
+        )
+
+    def publish_history_log(self, history_log: HistoryLog) -> None:
+        if self._setup_failed:
+            return
+
+        try:
+            timestamp_value = int(history_log["event_timestamp"].timestamp()) * 1000
+            message = to_json_str(history_log)
+            self._logs_client.put_log_events(
+                logGroupName=self.configuration.log_group_name,
+                logStreamName=self.configuration.log_stream_name,
+                logEvents=[
+                    InputLogEvent(
+                        timestamp=timestamp_value,
+                        message=message,
+                    )
+                ],
+            )
+        except Exception as ignored:
+            LOG.warning(
+                f"State Machine execution log event could not be published due to an error: '{ignored}'"
+            )
+
+    def setup(self):
+        # Create the log stream if one does not exist already.
+        try:
+            self._logs_client.create_log_stream(
```

Comment:
This is an interesting point that could provide resilience in case of deletion. However I have not observed a stream getting recreated in case of deletion during an execution. test_deleted_log_group simply check that the execution can take place if the logging configuration becomes faulty after creation (the group is deleted). However groups need to be created before the state machine is. I should add a test that checks if a stream deleted during execution is recreated immediately

Reply:
Deletion during execution is a super edge case we don't need to worry. It was more of an input for an alternative implementation.